### PR TITLE
CMakeToolchain: implement vs debugger path

### PR DESCRIPTION
**Pull request by jcar87**
Changelog: Feature: Set `CMAKE_VS_DEBUGGER_ENVIRONMENT` from CMakeToolchain to point to all binary directories when using Visual Studio. This negates the need to copy DLLs to launch executables from the Visual Studio IDE (requires CMake 3.27 or newer).

Docs: https://github.com/conan-io/docs/pull/3639

### CODE REVIEW

File: conan/tools/cmake/toolchain/blocks.py [hunk 1]
```diff
@@ -112,6 +112,53 @@ def context(self):
...
+    def context(self):
+        os_ = self._conanfile.settings.get_safe("os")
+        build_type = self._conanfile.settings.get_safe("build_type")
+
+        if (os_ and not "Windows" in os_) or not build_type:
+            return None
+        
+        generator = self._conanfile.conf.get("tools.cmake.cmaketoolchain:generator", default="Visual Studio")
+        if not "Visual Studio" in generator:
+            return None
```
Comment (memsharded) [inline conan/tools/cmake/toolchain/blocks.py]: You can check ``"Visual" not in self._toolchain.generator:``, the ``toolchain.generator`` already contains the generator defined

File: conan/tools/cmake/toolchain/blocks.py [hunk 2]
```diff
@@ -112,6 +112,53 @@ def context(self):
...
         return {"vs_runtimes": config_dict}
 
+class VSDebuggerEnvironment(Block):
+    template = textwrap.dedent("""
+        # Definition of CMAKE_VS_DEBUGGER_ENVIRONMENT
+        {% if vs_debugger_path %}
+        {% set genexpr = namespace(str='') %}
+        {% for config, value in vs_debugger_path.items() %}
+        {% set genexpr.str = genexpr.str +
+                                '$<$<CONFIG:' + config + '>:' + value|string + '>' %}
```
Comment (memsharded) [inline conan/tools/cmake/toolchain/blocks.py]: I prefer this logic in Python, is easier to maintain and debug IMHO (but good for me if you want to keep it)

File: conan/tools/cmake/toolchain/blocks.py [hunk 3]
```diff
@@ -112,6 +112,53 @@ def context(self):
...
+        
+        config_dict = {}
+        if os.path.exists(CONAN_TOOLCHAIN_FILENAME):
+            existing_include = load(CONAN_TOOLCHAIN_FILENAME)
+            vs_debugger_environment = re.search(r"set\(CMAKE_VS_DEBUGGER_ENVIRONMENT \"PATH=([^)]*)>;%PATH%\"\)",
+                                           existing_include)
+            if vs_debugger_environment:
+                capture = vs_debugger_environment.group(1)
+                matches = re.findall(r"\$<\$<CONFIG:([A-Za-z]*)>:(.*)", capture)
+                config_dict = dict(matches)
```
Comment (memsharded) [inline conan/tools/cmake/toolchain/blocks.py]: This is the part we might want to refactor, but that might be done later in another PR

**Review by memsharded — APPROVED** [review]
This is looking great

File: conan/tools/cmake/toolchain/blocks.py (hunk 2 above)
Comment (jcar87) [reply conan/tools/cmake/toolchain/blocks.py]: Done!

File: conan/tools/cmake/toolchain/blocks.py (hunk 1 above)
Comment (jcar87) [reply conan/tools/cmake/toolchain/blocks.py]: Done!

File: conan/tools/cmake/toolchain/blocks.py (hunk 3 above)
Comment (jcar87) [reply conan/tools/cmake/toolchain/blocks.py]: Removed the string generation from the jinja, and into python, although still very tricky to get right with the regexes. Fixed some issues when we have more than 2 configs, so it should be more robust.

IN a later PR I have some ideas for how to support multiconfig in a more straightforward way

**Review by memsharded — COMMENTED** [review]
Looking good, but tests are not passing, please check.

File: conan/tools/cmake/toolchain/blocks.py [hunk 4]
```diff
@@ -113,6 +113,54 @@ def context(self):
...
+
+    def context(self):
+        os_ = self._conanfile.settings.get_safe("os")
+        build_type = self._conanfile.settings.get_safe("build_type")
+
+        if (os_ and "Windows" not in os_) or not build_type:
+            return None
+
+        if "Visual" not in self._toolchain.generator:
+            return None
```
Comment (franramirez688) [inline conan/tools/cmake/toolchain/blocks.py:132]:
```suggestion
        if os_ != "Windows" or not build_type or "Visual" not in self._toolchain.generator:
            return None
```

File: conan/tools/cmake/toolchain/blocks.py [hunk 5]
```diff
@@ -113,6 +113,54 @@ def context(self):
...
+        test_deps = self._conanfile.dependencies.test.values()
+        bin_dirs = [p for dep in host_deps for p in dep.cpp_info.aggregated_components().bindirs]
+        test_bindirs = [p for dep in test_deps for p in dep.cpp_info.aggregated_components().bindirs]
+        bin_dirs.extend(test_bindirs)
+        bin_dirs = [p.replace("\\", "/") for p in bin_dirs]
+
+        bin_dirs = ";".join(bin_dirs)
+        config_dict[build_type] = bin_dirs
+        if all(value is '' for value in config_dict.values()):
+            return None
```
Comment (franramirez688) [inline conan/tools/cmake/toolchain/blocks.py]: Is this possible? Could we have this `bindirs = ["", "bin"]`? If not, I think it'd be better:

```suggestion
        if not bin_dirs:
            return None

        bin_dirs = ";".join(bin_dirs)
        config_dict[build_type] = bin_dirs
```

File: conans/test/functional/toolchains/cmake/test_cmake_toolchain.py [hunk 6]
```diff
@@ -189,6 +189,47 @@ def test_cmake_toolchain_without_build_type():
...
+    client.run("install --require=pkg/1.0 -s build_type=MinSizeRel -g CMakeToolchain --format=json")
+    minsizerel_graph = json.loads(client.stdout)
+    minsizerel_bindir = minsizerel_graph['graph']['nodes']['1']['cpp_info']['root']['bindirs'][0]
+    minsizerel_bindir = minsizerel_bindir.replace('\\', '/')
+
+    toolchain = client.load("conan_toolchain.cmake")
+    debugger_environment = f"PATH=$<$<CONFIG:Debug>:{debug_bindir}>" \
+                           f"$<$<CONFIG:Release>:{release_bindir}>" \
+                           f"$<$<CONFIG:MinSizeRel>:{minsizerel_bindir}>;%PATH%"
+    assert debugger_environment in toolchain
```
Comment (franramirez688) [inline conans/test/functional/toolchains/cmake/test_cmake_toolchain.py:230]: I'm missing a test checking the ways that get a `None`, i.e., the debugger variable is not appearing.

**Review by franramirez688 — COMMENTED** [review]
LGTM! 😁  I'm only missing a test checking when the VS variable is not appearing.

File: conan/tools/cmake/toolchain/blocks.py (hunk 4 above)
Comment (franramirez688) [reply conan/tools/cmake/toolchain/blocks.py:132]: Actually, it would not cover the WindowsCE and WindowsStore 👍

Comment (memsharded) [reply conan/tools/cmake/toolchain/blocks.py:132]: This won't cover the ``WindowsCE`` and ``WindoesStore`` that ``"Windows" in os_`` does cover.

File: conans/test/functional/toolchains/cmake/test_cmake_toolchain.py (hunk 6 above)
Comment (jcar87) [reply conans/test/functional/toolchains/cmake/test_cmake_toolchain.py:230]: added separate test `test_cmake_toolchain_cmake_vs_debugger_environment_not_needed` (which is also tested on other platforms, as well as Windows with a non-VS generator)

**Review by franramirez688 — APPROVED** [review]

**Review by memsharded — APPROVED** [review]
